File: bindings/rust/src/objectstore.rs
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) Jonathan D.A. Jewell <j.d.a.jewell@open.ac.uk>
//

//! Object Store types for the proven-servers ABI.
//!
//! Formally verified S3-compatible object store types.
//! Mirrors the Idris2 module `ObjectstoreABI.Types`.
//!
//! - `Operation` -- Object store operations.
//! - `StorageClass` -- Object storage classes.
//! - `Acl` -- Object ACL policies.
//! - `ErrorCode` -- Object store error codes.
//! - `SessionState` -- Object store session states.
//!
//! All discriminant values match the Idris2 ABI tag definitions exactly.

use std::fmt;

// ===========================================================================
// Object Store Constants
// ===========================================================================

/// Standard MinIO/S3 port.
pub const OBJECTSTORE_PORT: u16 = 9000;

// ===========================================================================
// Operation (tags 0-11)
// ===========================================================================

/// Object store operations.
///
/// Matches `Operation` in `ObjectstoreABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum Operation {
    /// PutObject (tag 0).
    PutObject = 0,
    /// GetObject (tag 1).
    GetObject = 1,
    /// DeleteObject (tag 2).
    DeleteObject = 2,
    /// ListObjects (tag 3).
    ListObjects = 3,
    /// HeadObject (tag 4).
    HeadObject = 4,
    /// CopyObject (tag 5).
    CopyObject = 5,
    /// CreateBucket (tag 6).
    CreateBucket = 6,
    /// DeleteBucket (tag 7).
    DeleteBucket = 7,
    /// ListBuckets (tag 8).
    ListBuckets = 8,
    /// InitMultipartUpload (tag 9).
    InitMultipartUpload = 9,
    /// UploadPart (tag 10).
    UploadPart = 10,
    /// CompleteMultipartUpload (tag 11).
    CompleteMultipartUpload = 11,
}

impl Operation {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::PutObject),
            1 => Some(Self::GetObject),
            2 => Some(Self::DeleteObject),
            3 => Some(Self::ListObjects),
            4 => Some(Self::HeadObject),
            5 => Some(Self::CopyObject),
            6 => Some(Self::CreateBucket),
            7 => Some(Self::DeleteBucket),
            8 => Some(Self::ListBuckets),
            9 => Some(Self::InitMultipartUpload),
            10 => Some(Self::UploadPart),
            11 => Some(Self::CompleteMultipartUpload),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// Whether this is a bucket-level operation.
    pub fn is_bucket_op(self) -> bool {
        matches!(self, Self::CreateBucket | Self::DeleteBucket | Self::ListBuckets)
    }

    /// Whether this is a multipart upload operation.
    pub fn is_multipart(self) -> bool {
        matches!(self, Self::InitMultipartUpload | Self::UploadPart | Self::CompleteMultipartUpload)
    }

    /// All variants of this type.
    pub const ALL: [Operation; 12] = [
        Self::PutObject, Self::GetObject, Self::DeleteObject, Self::ListObjects, Self::HeadObject, Self::CopyObject, Self::CreateBucket, Self::DeleteBucket, Self::ListBuckets, Self::InitMultipartUpload, Self::UploadPart, Self::CompleteMultipartUpload,
    ];
}

impl fmt::Display for Operation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// StorageClass (tags 0-4)
// ===========================================================================

/// Object storage classes.
///
/// Matches `StorageClass` in `ObjectstoreABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum StorageClass {
    /// Standard (tag 0).
    Standard = 0,
    /// InfrequentAccess (tag 1).
    InfrequentAccess = 1,
    /// Glacier (tag 2).
    Glacier = 2,
    /// DeepArchive (tag 3).
    DeepArchive = 3,
    /// OneZone (tag 4).
    OneZone = 4,
}

impl StorageClass {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Standard),
            1 => Some(Self::InfrequentAccess),
            2 => Some(Self::Glacier),
            3 => Some(Self::DeepArchive),
            4 => Some(Self::OneZone),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [StorageClass; 5] = [
        Self::Standard, Self::InfrequentAccess, Self::Glacier, Self::DeepArchive, Self::OneZone,
    ];
}

impl fmt::Display for StorageClass {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// Acl (tags 0-3)
// ===========================================================================

/// Object ACL policies.
///
/// Matches `Acl` in `ObjectstoreABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum Acl {
    /// Private (tag 0).
    Private = 0,
    /// PublicRead (tag 1).
    PublicRead = 1,
    /// PublicReadWrite (tag 2).
    PublicReadWrite = 2,
    /// AuthenticatedRead (tag 3).
    AuthenticatedRead = 3,
}

impl Acl {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Private),
            1 => Some(Self::PublicRead),
            2 => Some(Self::PublicReadWrite),
            3 => Some(Self::AuthenticatedRead),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [Acl; 4] = [
        Self::Private, Self::PublicRead, Self::PublicReadWrite, Self::AuthenticatedRead,
    ];
}

impl fmt::Display for Acl {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// ErrorCode (tags 0-7)
// ===========================================================================

/// Object store error codes.
///
/// Matches `ErrorCode` in `ObjectstoreABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum ErrorCode {
    /// NoSuchBucket (tag 0).
    NoSuchBucket = 0,
    /// NoSuchKey (tag 1).
    NoSuchKey = 1,
    /// BucketAlreadyExists (tag 2).
    BucketAlreadyExists = 2,
    /// BucketNotEmpty (tag 3).
    BucketNotEmpty = 3,
    /// AccessDenied (tag 4).
    AccessDenied = 4,
    /// EntityTooLarge (tag 5).
    EntityTooLarge = 5,
    /// InvalidPart (tag 6).
    InvalidPart = 6,
    /// IncompleteBody (tag 7).
    IncompleteBody = 7,
}

impl ErrorCode {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::NoSuchBucket),
            1 => Some(Self::NoSuchKey),
            2 => Some(Self::BucketAlreadyExists),
            3 => Some(Self::BucketNotEmpty),
            4 => Some(Self::AccessDenied),
            5 => Some(Self::EntityTooLarge),
            6 => Some(Self::InvalidPart),
            7 => Some(Self::IncompleteBody),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [ErrorCode; 8] = [
        Self::NoSuchBucket, Self::NoSuchKey, Self::BucketAlreadyExists, Self::BucketNotEmpty, Self::AccessDenied, Self::EntityTooLarge, Self::InvalidPart, Self::IncompleteBody,
    ];
}

impl fmt::Display for ErrorCode {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// SessionState (tags 0-4)
// ===========================================================================

/// Object store session states.
///
/// Matches `SessionState` in `ObjectstoreABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum SessionState {
    /// Idle (tag 0).
    Idle = 0,
    /// Ready (tag 1).
    Ready = 1,
    /// BucketActive (tag 2).
    BucketActive = 2,
    /// Uploading (tag 3).
    Uploading = 3,
    /// Closing (tag 4).
    Closing = 4,
}

impl SessionState {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Idle),
            1 => Some(Self::Ready),
            2 => Some(Self::BucketActive),
            3 => Some(Self::Uploading),
            4 => Some(Self::Closing),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [SessionState; 5] = [
        Self::Idle, Self::Ready, Self::BucketActive, Self::Uploading, Self::Closing,
    ];
}

impl fmt::Display for SessionState {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// Tests
// ===========================================================================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn operation_roundtrip() {
        for v in Operation::ALL {
            let tag = v.to_tag();
            let decoded = Operation::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(Operation::from_tag(12).is_none());
    }

    #[test]
    fn storage_class_roundtrip() {
        for v in StorageClass::ALL {
            let tag = v.to_tag();
            let decoded = StorageClass::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(StorageClass::from_tag(5).is_none());
    }

    #[test]
    fn acl_roundtrip() {
        for v in Acl::ALL {
            let tag = v.to_tag();
            let decoded = Acl::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(Acl::from_tag(4).is_none());
    }

    #[test]
    fn error_code_roundtrip() {
        for v in ErrorCode::ALL {
            let tag = v.to_tag();
            let decoded = ErrorCode::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(ErrorCode::from_tag(8).is_none());
    }

    #[test]
    fn session_state_roundtrip() {
        for v in SessionState::ALL {
            let tag = v.to_tag();
            let decoded = SessionState::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(SessionState::from_tag(5).is_none());
    }

    #[test]
    fn constants_match_idris() {
        assert_eq!(OBJECTSTORE_PORT, 9000);
    }

}

File: protocols/proven-objectstore/ffi/zig/src/objectstore.zig
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) Jonathan D.A. Jewell <j.d.a.jewell@open.ac.uk>
//
// objectstore.zig -- Zig FFI implementation of proven-objectstore.
//
// Implements an S3-compatible object store state machine with:
//   - 64-slot mutex-protected session pool
//   - Bucket registry (max 32 buckets per session)
//   - Object metadata tracking per bucket (max 256 objects total)
//   - Multipart upload state tracking
//   - ACL enforcement per bucket
//   - Storage class assignment per object
//   - Thread-safe via per-pool mutex
//
// All exported functions use C calling convention (callconv(.c)) and
// communicate state via u8 tags matching ObjectstoreABI.Types.idr exactly.

const std = @import("std");

// =========================================================================
// Enums (matching ObjectstoreABI.Types.idr tag assignments)
// =========================================================================

/// S3-compatible API operations (ABI tags 0-11).
pub const Operation = enum(u8) {
    put_object = 0,
    get_object = 1,
    delete_object = 2,
    list_objects = 3,
    head_object = 4,
    copy_object = 5,
    create_bucket = 6,
    delete_bucket = 7,
    list_buckets = 8,
    init_multipart_upload = 9,
    upload_part = 10,
    complete_multipart_upload = 11,
};

/// Storage classes (ABI tags 0-4).
pub const StorageClass = enum(u8) {
    standard = 0,
    infrequent_access = 1,
    glacier = 2,
    deep_archive = 3,
    one_zone = 4,
};

/// Access control list presets (ABI tags 0-3).
pub const ACL = enum(u8) {
    private = 0,
    public_read = 1,
    public_read_write = 2,
    authenticated_read = 3,
};

/// Error codes (ABI tags 0-7).
pub const ObjErrorCode = enum(u8) {
    no_such_bucket = 0,
    no_such_key = 1,
    bucket_already_exists = 2,
    bucket_not_empty = 3,
    access_denied = 4,
    entity_too_large = 5,
    invalid_part = 6,
    incomplete_body = 7,
};

/// Session lifecycle states (ABI tags 0-4).
pub const SessionState = enum(u8) {
    idle = 0,
    ready = 1,
    bucket_active = 2,
    uploading = 3,
    closing = 4,
};

// =========================================================================
// Internal data structures
// =========================================================================

const MAX_SESSIONS: usize = 64;
const MAX_BUCKETS: usize = 32;
const MAX_OBJECTS: usize = 256;
const MAX_NAME_LEN: usize = 256;
const MAX_REGION_LEN: usize = 64;

/// A bucket entry.
const BucketEntry = struct {
    name: [MAX_NAME_LEN]u8,
    name_len: u32,
    acl: ACL,
    object_count: u32,
    active: bool,
};

/// An object metadata entry.
const ObjectEntry = struct {
    key: [MAX_NAME_LEN]u8,
    key_len: u32,
    bucket_idx: u32,
    storage_class: StorageClass,
    size: u64,
    active: bool,
};

const empty_bucket: BucketEntry = .{
    .name = [_]u8{0} ** MAX_NAME_LEN,
    .name_len = 0,
    .acl = .private,
    .object_count = 0,
    .active = false,
};

const empty_object: ObjectEntry = .{
    .key = [_]u8{0} ** MAX_NAME_LEN,
    .key_len = 0,
    .bucket_idx = 0,
    .storage_class = .standard,
    .size = 0,
    .active = false,
};

/// An object store session.
const Session = struct {
    state: SessionState,
    region: [MAX_REGION_LEN]u8,
    region_len: u32,
    buckets: [MAX_BUCKETS]BucketEntry,
    bucket_count: u32,
    objects: [MAX_OBJECTS]ObjectEntry,
    object_count: u32,
    /// Index of the currently selected bucket (-1 if none).
    selected_bucket: i32,
    /// Whether a multipart upload is in progress.
    multipart_key: [MAX_NAME_LEN]u8,
    multipart_key_len: u32,
    multipart_parts: u32,
    active: bool,
};

const empty_session: Session = .{
    .state = .idle,
    .region = [_]u8{0} ** MAX_REGION_LEN,
    .region_len = 0,
    .buckets = [_]BucketEntry{empty_bucket} ** MAX_BUCKETS,
    .bucket_count = 0,
    .objects = [_]ObjectEntry{empty_object} ** MAX_OBJECTS,
    .object_count = 0,
    .selected_bucket = -1,
    .multipart_key = [_]u8{0} ** MAX_NAME_LEN,
    .multipart_key_len = 0,
    .multipart_parts = 0,
    .active = false,
};

// =========================================================================
// Global state
// =========================================================================

var sessions: [MAX_SESSIONS]Session = [_]Session{empty_session} ** MAX_SESSIONS;
var mutex: std.Thread.Mutex = .{};

// =========================================================================
// Internal helpers
// =========================================================================

fn validSlot(slot: c_int) ?usize {
    if (slot < 0 or slot >= MAX_SESSIONS) return null;
    const idx: usize = @intCast(slot);
    if (!sessions[idx].active) return null;
    return idx;
}

fn findBucket(idx: usize, name: []const u8) ?usize {
    for (&sessions[idx].buckets, 0..) |*b, i| {
        if (b.active and b.name_len == name.len and
            std.mem.eql(u8, b.name[0..b.name_len], name))
        {
            return i;
        }
    }
    return null;
}

// =========================================================================
// Exported C ABI functions
// =========================================================================

pub export fn objectstore_abi_version() callconv(.c) u32 {
    return 1;
}

/// Create a new object store session. Returns slot (>=0) or -1.
/// Transitions: Idle -> Ready.
pub export fn objectstore_create(
    region_ptr: [*]const u8,
    region_len: u32,
) callconv(.c) c_int {
    mutex.lock();
    defer mutex.unlock();

    if (region_len == 0 or region_len > MAX_REGION_LEN) return -1;

    for (&sessions, 0..) |*s, i| {
        if (!s.active) {
            s.* = empty_session;
            @memcpy(s.region[0..region_len], region_ptr[0..region_len]);
            s.region_len = region_len;
            s.state = .ready;
            s.active = true;
            return @intCast(i);
        }
    }
    return -1;
}

pub export fn objectstore_destroy(slot: c_int) callconv(.c) void {
    mutex.lock();
    defer mutex.unlock();
    if (slot < 0 or slot >= MAX_SESSIONS) return;
    sessions[@intCast(slot)] = empty_session;
}

pub export fn objectstore_state(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return @intFromEnum(sessions[idx].state);
}

/// Create a bucket. Returns 0 on success, 1 on rejection.
pub export fn objectstore_create_bucket(
    slot: c_int,
    name_ptr: [*]const u8,
    name_len: u32,
    acl: u8,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .ready and sessions[idx].state != .bucket_active) return 1;
    if (name_len == 0 or name_len > MAX_NAME_LEN) return 1;
    if (acl > 3) return 1;

    const name = name_ptr[0..name_len];
    if (findBucket(idx, name) != null) return 1; // already exists

    for (&sessions[idx].buckets) |*b| {
        if (!b.active) {
            @memcpy(b.name[0..name_len], name);
            b.name_len = name_len;
            b.acl = @enumFromInt(acl);
            b.object_count = 0;
            b.active = true;
            sessions[idx].bucket_count += 1;
            return 0;
        }
    }
    return 1;
}

/// Delete a bucket. Returns 0 on success, 1 on rejection.
/// Rejects if bucket has objects (BucketNotEmpty).
pub export fn objectstore_delete_bucket(
    slot: c_int,
    name_ptr: [*]const u8,
    name_len: u32,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .ready and sessions[idx].state != .bucket_active) return 1;
    if (name_len == 0 or name_len > MAX_NAME_LEN) return 1;

    const name = name_ptr[0..name_len];
    const bi = findBucket(idx, name) orelse return 1;

    if (sessions[idx].buckets[bi].object_count > 0) return 1; // not empty

    sessions[idx].buckets[bi].active = false;
    sessions[idx].bucket_count -= 1;

    // If this was the selected bucket, deselect
    if (sessions[idx].selected_bucket == @as(i32, @intCast(bi))) {
        sessions[idx].selected_bucket = -1;
        if (sessions[idx].state == .bucket_active) {
            sessions[idx].state = .ready;
        }
    }
    return 0;
}

/// Select a bucket for operations. Returns 0 on success, 1 on rejection.
/// Transitions: Ready -> BucketActive.
pub export fn objectstore_select_bucket(
    slot: c_int,
    name_ptr: [*]const u8,
    name_len: u32,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .ready and sessions[idx].state != .bucket_active) return 1;
    if (name_len == 0 or name_len > MAX_NAME_LEN) return 1;

    const name = name_ptr[0..name_len];
    const bi = findBucket(idx, name) orelse return 1;

    sessions[idx].selected_bucket = @intCast(bi);
    sessions[idx].state = .bucket_active;
    return 0;
}

/// Put an object into the selected bucket. Returns 0 on success, 1 on rejection.
pub export fn objectstore_put_object(
    slot: c_int,
    key_ptr: [*]const u8,
    key_len: u32,
    body_ptr: [*]const u8,
    body_len: u32,
    storage_class: u8,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    _ = body_ptr;

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .bucket_active) return 1;
    if (key_len == 0 or key_len > MAX_NAME_LEN) return 1;
    if (storage_class > 4) return 1;
    if (sessions[idx].selected_bucket < 0) return 1;

    const bi: usize = @intCast(sessions[idx].selected_bucket);

    // Find or create object slot
    const key = key_ptr[0..key_len];
    for (&sessions[idx].objects) |*obj| {
        if (obj.active and obj.bucket_idx == bi and
            obj.key_len == key_len and
            std.mem.eql(u8, obj.key[0..obj.key_len], key))
        {
            // Overwrite existing
            obj.storage_class = @enumFromInt(storage_class);
            obj.size = body_len;
            return 0;
        }
    }

    // New object
    for (&sessions[idx].objects) |*obj| {
        if (!obj.active) {
            @memcpy(obj.key[0..key_len], key);
            obj.key_len = key_len;
            obj.bucket_idx = @intCast(bi);
            obj.storage_class = @enumFromInt(storage_class);
            obj.size = body_len;
            obj.active = true;
            sessions[idx].object_count += 1;
            sessions[idx].buckets[bi].object_count += 1;
            return 0;
        }
    }
    return 1;
}

/// Check if an object exists (head). Returns 1 if exists, 0 if not.
pub export fn objectstore_head_object(
    slot: c_int,
    key_ptr: [*]const u8,
    key_len: u32,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 0;
    if (sessions[idx].state != .bucket_active) return 0;
    if (key_len == 0 or key_len > MAX_NAME_LEN) return 0;
    if (sessions[idx].selected_bucket < 0) return 0;

    const bi: usize = @intCast(sessions[idx].selected_bucket);
    const key = key_ptr[0..key_len];

    for (&sessions[idx].objects) |*obj| {
        if (obj.active and obj.bucket_idx == bi and
            obj.key_len == key_len and
            std.mem.eql(u8, obj.key[0..obj.key_len], key))
        {
            return 1;
        }
    }
    return 0;
}

/// Get an object (simulated). Returns 0 on success, 1 if not found.
pub export fn objectstore_get_object(
    slot: c_int,
    key_ptr: [*]const u8,
    key_len: u32,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .bucket_active) return 1;
    if (key_len == 0 or key_len > MAX_NAME_LEN) return 1;
    if (sessions[idx].selected_bucket < 0) return 1;

    const bi: usize = @intCast(sessions[idx].selected_bucket);
    const key = key_ptr[0..key_len];

    for (&sessions[idx].objects) |*obj| {
        if (obj.active and obj.bucket_idx == bi and
            obj.key_len == key_len and
            std.mem.eql(u8, obj.key[0..obj.key_len], key))
        {
            return 0;
        }
    }
    return 1;
}

/// Delete an object. Returns 0 on success, 1 on rejection.
pub export fn objectstore_delete_object(
    slot: c_int,
    key_ptr: [*]const u8,
    key_len: u32,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .bucket_active) return 1;
    if (key_len == 0 or key_len > MAX_NAME_LEN) return 1;
    if (sessions[idx].selected_bucket < 0) return 1;

    const bi: usize = @intCast(sessions[idx].selected_bucket);
    const key = key_ptr[0..key_len];

    for (&sessions[idx].objects) |*obj| {
        if (obj.active and obj.bucket_idx == bi and
            obj.key_len == key_len and
            std.mem.eql(u8, obj.key[0..obj.key_len], key))
        {
            obj.active = false;
            sessions[idx].object_count -= 1;
            sessions[idx].buckets[bi].object_count -= 1;
            return 0;
        }
    }
    return 1;
}

/// Initiate a multipart upload. Returns 0 on success, 1 on rejection.
/// Transitions: BucketActive -> Uploading.
pub export fn objectstore_init_multipart(
    slot: c_int,
    key_ptr: [*]const u8,
    key_len: u32,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .bucket_active) return 1;
    if (key_len == 0 or key_len > MAX_NAME_LEN) return 1;

    @memcpy(sessions[idx].multipart_key[0..key_len], key_ptr[0..key_len]);
    sessions[idx].multipart_key_len = key_len;
    sessions[idx].multipart_parts = 0;
    sessions[idx].state = .uploading;
    return 0;
}

/// Complete a multipart upload. Returns 0 on success, 1 on rejection.
/// Transitions: Uploading -> BucketActive.
pub export fn objectstore_complete_multipart(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .uploading) return 1;

    sessions[idx].state = .bucket_active;
    sessions[idx].multipart_key_len = 0;
    sessions[idx].multipart_parts = 0;
    return 0;
}

/// Returns the number of buckets.
pub export fn objectstore_bucket_count(slot: c_int) callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return sessions[idx].bucket_count;
}

/// Close the session. Returns 0 on success, 1 on rejection.
pub export fn objectstore_close(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    const state = sessions[idx].state;
    if (state == .ready or state == .bucket_active) {
        sessions[idx].state = .closing;
        return 0;
    }
    return 1;
}

/// Complete cleanup. Returns 0 on success, 1 on rejection.
/// Transitions: Closing -> Idle.
pub export fn objectstore_cleanup(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .closing) return 1;

    sessions[idx].state = .idle;
    sessions[idx].buckets = [_]BucketEntry{empty_bucket} ** MAX_BUCKETS;
    sessions[idx].bucket_count = 0;
    sessions[idx].objects = [_]ObjectEntry{empty_object} ** MAX_OBJECTS;
    sessions[idx].object_count = 0;
    sessions[idx].selected_bucket = -1;
    return 0;
}

// --- pool size guard (audit S5: prevent oversized-global stack overflow) ---
comptime {
    if (@sizeOf(@TypeOf(sessions)) > 16 * 1024 * 1024)
        @compileError("pool 'sessions' exceeds the 16 MiB budget; heap-allocate or shrink (see audits/proof-panic-attack-2026-06-23.md)");
}
